File: src/backend/src/main/java/com/ruleengine/service/RuleParser.java
package com.ruleengine.service;

import com.ruleengine.model.ASTNode;
import org.springframework.stereotype.Service;

@Service
public class RuleParser {
    private int pos = 0;
    private String input = "";

    public ASTNode parse(String ruleString) {
        pos = 0;
        input = ruleString.trim();
        return parseExpression();
    }

    private ASTNode parseExpression() {
        ASTNode node = parseTerm();

        while (pos < input.length()) {
            String operator = consumeOperator();
            if (operator == null) break;

            ASTNode right = parseTerm();
            ASTNode newNode = new ASTNode();
            newNode.setType("operator");
            newNode.setOperator(operator);
            newNode.setLeft(node);
            newNode.setRight(right);
            node = newNode;
        }

        return node;
    }

    private ASTNode parseTerm() {
        consumeWhitespace();

        if (input.charAt(pos) == '(') {
            pos++; // Skip '('
            ASTNode node = parseExpression();
            pos++; // Skip ')'
            return node;
        }

        String field = consumeIdentifier();
        String operator = consumeOperator();
        String value = consumeValue();

        ASTNode node = new ASTNode();
        node.setType("operand");
        node.setField(field);
        node.setOperator(operator);
        node.setValue(value);
        return node;
    }

    private void consumeWhitespace() {
        while (pos < input.length() && Character.isWhitespace(input.charAt(pos))) {
            pos++;
        }
    }

    private String consumeIdentifier() {
        consumeWhitespace();
        StringBuilder identifier = new StringBuilder();
        
        while (pos < input.length() && (Character.isLetterOrDigit(input.charAt(pos)) || input.charAt(pos) == '_')) {
            identifier.append(input.charAt(pos));
            pos++;
        }
        
        return identifier.toString();
    }

    private String consumeOperator() {
        consumeWhitespace();
        String[] operators = {"AND", "OR", ">", "<", "=", ">=", "<="};
        
        for (String op : operators) {
            if (input.startsWith(op, pos)) {
                pos += op.length();
                return op;
            }
        }
        
        return null;
    }

    private String consumeValue() {
        consumeWhitespace();
        StringBuilder value = new StringBuilder();
        
        if (input.charAt(pos) == '\'') {
            pos++; // Skip opening quote
            while (pos < input.length() && input.charAt(pos) != '\'') {
                value.append(input.charAt(pos));
                pos++;
            }
            pos++; // Skip closing quote
        } else {
            while (pos < input.length() && Character.isDigit(input.charAt(pos))) {
                value.append(input.charAt(pos));
                pos++;
            }
        }
        
        return value.toString();
    }
}

File: src/utils/ruleParser.ts
import { ASTNode, Operator } from '../types/ast';

export class RuleParser {
  private pos = 0;
  private input = '';

  public parse(ruleString: string): ASTNode {
    this.pos = 0;
    this.input = ruleString.trim();
    return this.parseExpression();
  }

  private parseExpression(): ASTNode {
    let node = this.parseTerm();

    while (this.pos < this.input.length) {
      const operator = this.consumeOperator();
      if (!operator) break;

      const right = this.parseTerm();
      node = {
        type: 'operator',
        operator: operator as Operator,
        left: node,
        right
      };
    }

    return node;
  }

  private parseTerm(): ASTNode {
    this.consumeWhitespace();

    if (this.input[this.pos] === '(') {
      this.pos++; // Skip '('
      const node = this.parseExpression();
      this.pos++; // Skip ')'
      return node;
    }

    // Parse field, operator, and value
    const field = this.consumeIdentifier();
    const operator = this.consumeOperator();
    const value = this.consumeValue();

    return {
      type: 'operand',
      field,
      operator: operator as Operator,
      value
    };
  }

  private consumeWhitespace(): void {
    while (this.pos < this.input.length && /\s/.test(this.input[this.pos])) {
      this.pos++;
    }
  }

  private consumeIdentifier(): string {
    this.consumeWhitespace();
    let identifier = '';
    
    while (this.pos < this.input.length && /[a-zA-Z_]/.test(this.input[this.pos])) {
      identifier += this.input[this.pos];
      this.pos++;
    }

    return identifier;
  }

  private consumeOperator(): string | null {
    this.consumeWhitespace();
    const operators = ['AND', 'OR', '>', '<', '=', '>=', '<='];
    
    for (const op of operators) {
      if (this.input.substr(this.pos, op.length) === op) {
        this.pos += op.length;
        return op;
      }
    }
    
    return null;
  }

  private consumeValue(): string | number {
    this.consumeWhitespace();
    let value = '';
    
    if (this.input[this.pos] === "'") {
      this.pos++; // Skip opening quote
      while (this.pos < this.input.length && this.input[this.pos] !== "'") {
        value += this.input[this.pos];
        this.pos++;
      }
      this.pos++; // Skip closing quote
      return value;
    }
    
    while (this.pos < this.input.length && /[0-9]/.test(this.input[this.pos])) {
      value += this.input[this.pos];
      this.pos++;
    }
    
    return Number(value);
  }
}